cmd/dev.go

```diff
@@ -0,0 +1,677 @@
+			continue
+		}
+
+		microPort, err := parsePort(portFile)
+		if err != nil {
+			stoppedMicros = append(stoppedMicros, micro)
+			continue
+		}
+
+		if isRunning := isPortActive(microPort); !isRunning {
+			stoppedMicros = append(stoppedMicros, micro)
+			continue
+		}
+	}
+
+	commands := make([]*exec.Cmd, 0, len(stoppedMicros))
+	startPort := proxyPort + 1
+	for _, micro := range stoppedMicros {
+		freePort, err := GetFreePort(startPort)
+		if err != nil {
+			return err
+		}
+
+		command, err := microCommand(micro, "", projectDir, projectKey, freePort)
+		if err != nil {
+			return err
+		}
+
+		portFile := path.Join(routeDir, fmt.Sprintf("%s.port", micro.Name))
+		if err := writePortFile(portFile, freePort); err != nil {
+			return err
+		}
+		defer os.Remove(portFile)
+
+		commands = append(commands, command)
+		startPort = freePort + 1
+	}
+
+	proxy, err := proxyFromDir(spacefile.Micros, routeDir)
+	if err != nil {
+		return err
+	}
+
+	addr := fmt.Sprintf("localhost:%d", proxyPort)
+	server := http.Server{
+		Addr:    addr,
+		Handler: proxy,
+	}
+
+	wg := sync.WaitGroup{}
+
+	for _, command := range commands {
+		wg.Add(1)
+		go func(command *exec.Cmd) {
+			defer wg.Done()
+			command.Run()
+		}(command)
+	}
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		logger.Println("proxy listening on", addr)
+		err := server.ListenAndServe()
+		if err != nil && err != http.ErrServerClosed {
+			logger.Println("proxy error", err)
+		}
+	}()
+
+	go func() {
+		sigs := make(chan os.Signal, 1)
+		signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+		<-sigs
+		logger.Println("shutting down all commands")
+		for _, command := range commands {
+			command.Process.Signal(syscall.SIGTERM)
+		}
+		server.Shutdown(context.Background())
+	}()
+
+	if open, _ := cmd.Flags().GetBool("open"); open {
+		browser.OpenURL(fmt.Sprintf("http://%s", addr))
+	}
+
+	wg.Wait()
+
+	// Wait a bit for all logs to be printed
+	time.Sleep(1 * time.Second)
+
+	return nil
+}
+
+func writePortFile(filepath string, port int) error {
+	portDir := path.Dir(filepath)
+	if _, err := os.Stat(portDir); os.IsNotExist(err) {
+		if err := os.MkdirAll(portDir, 0755); err != nil {
+			return err
+		}
+	}
+
+	return os.WriteFile(filepath, []byte(fmt.Sprintf("%d", port)), 0644)
+}
+
+func proxyFromDir(micros []*shared.Micro, routeDir string) (*proxy.ReverseProxy, error) {
+	routes := make([]proxy.ProxyRoute, 0)
+	for _, micro := range micros {
+		portFile := path.Join(routeDir, fmt.Sprintf("%s.port", micro.Name))
+		if _, err := os.Stat(portFile); err != nil {
+			continue
+		}
+
+		microPort, err := parsePort(portFile)
+		if err != nil {
+			continue
+		}
+
+		target, _ := url.Parse(fmt.Sprintf("http://localhost:%d", microPort))
+
+		logger.Println("proxying", micro.Prefix(), "to", target.String())
+		routes = append(routes, proxy.ProxyRoute{
+			Prefix: micro.Prefix(),
+			Target: target,
+		})
+	}
+
+	return proxy.NewReverseProxy(routes), nil
```

Comment:
We could add a header to proxied requests so the Micro knows that a request has been proxied using `space dev`. The `Via` header is meant for proxies so we could use it here as well ([MDN docs on it](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Via))